pkg/controller/clone-controller.go

```diff
@@ -0,0 +1,270 @@
+			}
+		},
+	})
+
+	// Bind the pod SharedIndexInformer to the pod queue
+	c.podInformer.AddEventHandler(cache.ResourceEventHandlerFuncs{
+		AddFunc: func(obj interface{}) {
+			key, err := cache.MetaNamespaceKeyFunc(obj)
+			if err == nil {
+				c.podQueue.AddRateLimited(key)
+			}
+		},
+		UpdateFunc: func(oldObj, newObj interface{}) {
+			key, err := cache.MetaNamespaceKeyFunc(newObj)
+			if err == nil {
+				c.podQueue.AddRateLimited(key)
+			}
+		},
+	})
+
+	return c
+}
+
+func (c *CloneController) Run(threadiness int, stopCh <-chan struct{}) error {
+	defer func() {
+		c.pvcQueue.ShutDown()
+		c.podQueue.ShutDown()
+	}()
+	glog.V(Vadmin).Infoln("Starting clone controller Run loop")
+	if threadiness < 1 {
+		return errors.Errorf("expected >0 threads, got %d", threadiness)
+	}
+
+	if !cache.WaitForCacheSync(stopCh, c.pvcInformer.HasSynced) {
+		return errors.New("Timeout waiting for pvc cache sync")
+	}
+	if !cache.WaitForCacheSync(stopCh, c.podInformer.HasSynced) {
+		return errors.New("Timeout waiting for pod cache sync")
+	}
+	glog.V(Vdebug).Infoln("CloneController cache has synced")
+
+	for i := 0; i < threadiness; i++ {
+		go wait.Until(c.runPVCWorkers, time.Second, stopCh)
+		go wait.Until(c.runPodWorkers, time.Second, stopCh)
+	}
+	<-stopCh
+	return nil
+}
+
+func (c *CloneController) runPodWorkers() {
+	for c.ProcessNextPodItem() {
+	}
+}
+
+func (c *CloneController) runPVCWorkers() {
+	for c.ProcessNextPvcItem() {
+	}
+}
+
+// ProcessNextPodItem gets the next pod key from the queue and verifies that it was created by the
+// controller. If not the key is discarded; otherwise, the pod object is passed to processPodItem.
+// Note: pods are already filtered by label "host-assisted-cloning".
+func (c *CloneController) ProcessNextPodItem() bool {
+	key, shutdown := c.podQueue.Get()
+	if shutdown {
+		return false
+	}
+	defer c.podQueue.Done(key)
+	pod, err := c.podFromKey(key)
+	if err != nil {
+		c.forgetKey(key, fmt.Sprintf("ProcessNextPodItem: unable to get pod from key %v: %v", key, err))
+		return true
+	}
+	if !metav1.HasAnnotation(pod.ObjectMeta, AnnCloningCreatedBy) {
+		c.forgetKey(key, fmt.Sprintf("ProcessNextPodItem: pod %q does not have annotation %q", key, AnnCloningCreatedBy))
+		return true
+	}
+	glog.V(Vdebug).Infof("ProcessNextPodItem: next pod to process: %s\n", key)
+	err = c.processPodItem(pod)
+	if err != nil { // processPodItem errors may not have been logged so log here
+		glog.Errorf("error processing pod %q: %v", key, err)
+		return true
+	}
+	return c.forgetKey(key, fmt.Sprintf("ProcessNextPodItem: processing pod %q completed", key))
+}
+
+// processPodItem verifies that the passed in pod is genuine and, if so, annotates the Phase
+// of the pod in the PVC to indicate the status of the cloning process.
+func (c *CloneController) processPodItem(pod *v1.Pod) error {
+	// verify that this pod has the expected pvc name
+	var pvcKey string
+	for _, vol := range pod.Spec.Volumes {
+		if vol.Name == ImagePathName {
```

Comment:
move this volume name check to after you've verified that the pod here is the target pod (not source pod). No reason to loop here on the source pod.